nomt/src/page_walker.rs

```diff
@@ -69,12 +68,18 @@ pub enum Output {
     /// A new root node.
     ///
     /// This is always the output when no parent page is supplied to the walker.
-    Root(Node, HashMap<PageId, PageDiff>),
+    Root(Node, Vec<(PageId, PageDiff)>),
     /// Nodes to set in the bottom layer of the parent page, indexed by the position of the node
     /// to set.
     ///
     /// This is always the output when a parent page is supplied to the walker.
-    ChildPageRoots(Vec<(TriePosition, Node)>, HashMap<PageId, PageDiff>),
+    ChildPageRoots(Vec<(TriePosition, Node)>, Vec<(PageId, PageDiff)>),
+}
+
+struct StackItem {
+    page_id: PageId,
+    page: Page,
+    diff: PageDiff,
```

Comment:
Nit: the diff was added but it's not immediatelly obvious why. While it's easy to see why we have the two former items, it's not so easy with `diff`.

Reply:
IMO it's somewhat clear from the context of the page walker.

We walk the tree of pages from left to right, which means that our stack at any point is a kind of cross-section of the page tree. While pages are in that stack, they change. When they leave the stack, they won't be edited any more and we can save their diff.

(note: there are some edge cases where we pop a page off the stack and then right afterwards we either edit it or push it back for a single operation, and that is because of `write_leaf_children`, but in general this principle holds)